### PR TITLE
[BOUNTY] Directory page UI improvements

### PR DESCRIPTION
**Pull request by neatonk**
> BOUNTY details: https://github.com/ipfs/dir-index-html/issues/37
> This is a dependency of https://github.com/ipfs/go-ipfs/pull/7536

This PR implements the changes described in https://github.com/ipfs/dir-index-html/issues/37.

This is an incremental improvement over the previous version of the `dir-index.html` page and I suggest that a larger restructuring of the page should be considered after this PR is completed. Additional changes informed by a discussion of design constraints (e.g. file size and responsiveness) and developer workflow (e.g. build tools and minification) would lead to greater improvements.

Screen Shot 2020-07-07 at 09 24 38
Screen Shot 2020-07-07 at 09 24 45
Screen Shot 2020-07-07 at 09 25 00
Screen Shot 2020-07-07 at 09 25 09
Screen Shot 2020-07-07 at 09 26 23

Closes https://github.com/ipfs/dir-index-html/issues/37

### CODE REVIEW

**Comment by jessicaschilling** [comment]
Thanks, @neatonk! 

This looks good to me but I'd like to have a closer review by the go-ipfs core team. @aschmahmann -- do you have the time, or can you suggest another reviewer?

Also: @neatonk, can you please drop me a line? jessica at protocol dot ai.

**Comment by jessicaschilling** [comment]
Thanks, @neatonk!

This looks good to me but I'd like to have a closer review by the go-ipfs core team. @aschmahmann -- do you have the time, or can you suggest another reviewer?

**Comment by jessicaschilling** [comment]
@aschmahmann -- just to note, this behaves as intended (meets all the criteria in https://github.com/ipfs/dir-index-html/issues/37#issue-644155427 ) using the test server described [here](https://github.com/ipfs/dir-index-html#testing).

I'm afraid I can't explicitly confirm/deny that the test info gives us enough confidence to merge. Is this something that you or @Stebalien can investigate? Apologies for the inconvenience.

File: test/main.go [hunk 1]
```diff
@@ -13,7 +13,9 @@ const templateFile = "../dir-index.html"
 // Copied from go-ipfs/core/corehttp/gateway_indexPage.go
```
Comment (neatonk) [inline test/main.go:13]: I believe the updated `dir-index.html` template can be safely merged and be used upstream, after the type declarations below have been included. All of the new struct fields are optional and the updated `dir-index.html` template can be safely used without them. However, to see most of the UI improvements provided by of this PR, the new struct fields will need to be populated as in the code below.

Comment (aschmahmann) [reply test/main.go:13]: @jessicaschilling @neatonk it looks like the changes in this library are fine, but we can't reasonably merge this until the corresponding changes are made in the go-ipfs gateway to emit the new template struct fields.

@jessicaschilling Is implementing the changes to go-ipfs part of the bounty as well?

**Comment by jessicaschilling** [comment]
It wasn’t in the original bounty spec - however, @neatonk, is that something you could take on? We may be able to adjust the bounty as a result.

File: test/main.go (hunk 1 above)
Comment (jessicaschilling) [reply test/main.go:13]: It wasn’t in the original bounty spec - however, @neatonk, is that something you could take on? We may be able to adjust the bounty as a result.

Comment (neatonk) [reply test/main.go:13]: Yep, I can do that. 

Regarding the bounty, I think it would be best to either increase the payment to reflect the additional work needed or close it out for the work that was done. I am fine with either option.

File: dir-index.html [hunk 2]
```diff
@@ -39,36 +39,61 @@
...
+          <div class="{{iconFromExt .Name}}">&nbsp;</div>
+        </td>
+        <td>
+          <a href="{{ .Path | urlEscape }}">{{ .Name }}</a>
+        </td>
+        <td>
+          {{ if .Hash }}
+          <a class="d-flex" style="width:33vw; max-width:30em; max-width: max-content" href="/ipfs/{{ .Hash | urlEscape }}">
+            <span class="text-truncate" style="min-width: 2.5em">{{ slice .Hash 0 40 }}</span>
+            <span class="no-linebreak">{{ slice .Hash 40 }}</span>
```
[the remark this answers is not in the document]
Comment (jessicaschilling) [reply dir-index.html:88]: +1 to what @lidel suggests -- first four and last four are great. That's really all the user functionally needs to know anyway.

**Comment by neatonk** [comment]
@jessicaschilling Is dir-index-uncat.html actually in use? If not, I would prefer to leave it alone. It is not clear how the changes I made should be applied and for what purpose.

**Comment by jessicaschilling** [comment]
They are both in use, unfortunately. https://github.com/ipfs/dir-index-html#updating does include guidance to touch them both when updating, but this also could have been called out more explicitly in the issue; that's my fault.

**Comment by neatonk** [comment]
No worries and thanks for pointing that out. However, it looks to me like the two files are already out of sync. That will make it difficult for me to update `dir-index-uncat.html` to include the changes in this PR. 

Below are two possible solutions to this problem. I am in favor of **Option B** since it would be a better use of time and would not delay this PR.

**Option A**

First, tap someone who is familiar with previous changes to `dir-index.html` and ask them to update `dir-index-uncat.html` to reflect those changes and commit the result. Then I can rebase my branch to include that commit and update `dir-index-uncat.html` to reflect the changes in this PR.

**Option B**

Finalize this PR (and bounty) without updating `dir-index-uncat.html`. Plan to correct this later by building `dir-index.html` from `dir-index-uncat.html` (possibly via #6). This is essentially what I was alluding to when I mentioned "developer workflow" [above](https://github.com/ipfs/dir-index-html/pull/38#issue-445429192).

**Comment by jessicaschilling** [comment]
- Option B is fine. I'll need to rationalize the two files at a later date before we can release this, but if you're able to get everything working and tested for dir-index.html that's a start. 

Agreed that automating as in #6 is the best long-term solution, but we've not got anyone with the bandwidth for that at present.

File: dir-index.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (neatonk) [reply dir-index.html]: See ce0dcea

File: dir-index.html [hunk 3]
```diff
@@ -39,36 +40,60 @@
...
+        <td>
+          <a href="{{ .Path | urlEscape }}">{{ .Name }}</a>
+        </td>
+        <td class="no-linebreak">
+          {{ if .Hash }}
+          <a href="{{ $root.GatewayURL }}/ipfs/{{ .Hash | urlEscape }}">
+            {{ .ShortHash }}
+          </a>
+          {{ end }}
+        </td>
```
Comment (lidel) [inline dir-index.html:91]: @jessicaschilling thoughts on making CIDs monospace and use the same color as CID in the header to not distract from actual listing?

I think we could extend & reuse `ipfs-hash` class here – made a quick mockup:




(AFAIK this is not a part of the Bounty, so could be a separate PR)

Comment (jessicaschilling) [reply dir-index.html:91]: I'll do this in a separate PR - need to port changes over to dir-index-uncat anyway, so can do these in tandem.

**Comment by jessicaschilling** [comment]
Also branch `chore/reconcile-uncat`
